### PR TITLE
Add initial commutative algebra

### PR DESCRIPTION
**Pull request by felixwellen**
Add the initial commutative R-Algebra, which is just R itself and prove its universal property.

### CODE REVIEW

File: Cubical/Algebra/CommAlgebra/Base.agda [hunk 1]
```diff
@@ -160,6 +160,49 @@ module _ {R : CommRing ℓ} where
...
+      instance
+        _ = snd M
+        _ = snd N
+    makeCommAlgebraHom : (f : fst M → fst N)
+                           → (pres1 : f 1a ≡ 1a)
+                           → (pres+ : (x y : fst M) → f (x + y) ≡ f x + f y)
+                           → (pres· : (x y : fst M) → f (x · y) ≡ f x · f y)
+                           → (pres⋆ : (r : fst R) (x : fst M) → f (r ⋆ x) ≡ r ⋆ f x)
+                           → CommAlgebraHom M N
+    makeCommAlgebraHom f pres1 pres+ pres· pres⋆ = f , record
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Base.agda:175]: Please define using copatterns

File: Cubical/Algebra/CommAlgebra/Base.agda [hunk 2]
```diff
@@ -160,6 +160,49 @@ module _ {R : CommRing ℓ} where
...
+            }
+            where pres0 =
+                    f 0a                  ≡⟨ sym (+-rid _) ⟩
+                    f 0a + 0a             ≡⟨ cong (λ u → f 0a + u) (sym (+-rinv (f 0a))) ⟩
+                    f 0a + (f 0a - f 0a)  ≡⟨ +-assoc (f 0a) (f 0a) (- f 0a) ⟩
+                    (f 0a + f 0a) - f 0a  ≡⟨ cong (λ u → u - f 0a) (sym (pres+ 0a 0a)) ⟩
+                    f (0a + 0a) - f 0a    ≡⟨ cong (λ u → f u - f 0a) (+-lid 0a) ⟩
+                    f 0a - f 0a           ≡⟨ +-rinv (f 0a) ⟩
+                    0a ∎
+    isPropIsCommAlgebraHom : (f : fst M → fst N) → isProp (IsCommAlgebraHom (snd M) f (snd N))
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Base.agda:199]: Add line break above this

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 3]
```diff
@@ -0,0 +1,76 @@
+{-# OPTIONS --cubical --safe --no-import-sorts #-}
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:1]: You don't need all of these flags now. `--safe` should be enough

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+module Cubical.Algebra.CommAlgebra.Instances.Initial where
+
+open import Cubical.Foundations.Everything
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:4]: Please don't import `Everything` files

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+private
+  variable
+    ℓ : Level
+
+module CommAlgebraExamples ((R , str) : CommRing ℓ) where
+
+  initial : CommAlgebra (R , str) ℓ
+  initial =
+    let open CommRingStr str
+    in  (R , commalgebrastr _ _ _ _ _ (λ r x → r · x)
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:24]: Please use copatterns

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      _ = snd A
+      _ : CommAlgebraStr (R , str) R
+      _ = snd initial
+
+    _*_ : R → (fst A) → (fst A)
+    r * a = CommAlgebraStr._⋆_ (snd A) r a
+
+    initialMap : CommAlgebraHom initial A
+    initialMap =
+      makeCommAlgebraHom {M = initial} {N = A}
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:46]: Copatterns

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+open import Cubical.Algebra.Algebra.Base using (IsAlgebraHom)
+open import Cubical.Algebra.CommAlgebra.Base
+
+private
+  variable
+    ℓ : Level
+
+module CommAlgebraExamples ((R , str) : CommRing ℓ) where
+
+  initial : CommAlgebra (R , str) ℓ
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:21]: Call this `initialCommAlgebra`? Same changes apply to other names in this module

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+           (isPropIsCommAlgebraHom {M = initial} {N = A})
+             λ i x →
+               ((fst f) x                              ≡⟨ cong (fst f) (sym (·Rid _)) ⟩
+               fst f (x · 1a)                          ≡⟨ pres⋆ x 1a ⟩
+               CommAlgebraStr._⋆_ (snd A) x (fst f 1a) ≡⟨ cong
+                                                           (λ u → (snd A CommAlgebraStr.⋆ x) u)
+                                                           pres1 ⟩
+               (CommAlgebraStr._⋆_ (snd A) x 1a) ∎) i
+
+    isInitial : CommAlgebraHom initial A ≡ Unit*
```
Comment (mortberg) [inline Cubical/Algebra/CommAlgebra/Instances/Initial.agda:71]: Maybe give iso version a name?

**Review by mortberg — CHANGES_REQUESTED** [review]
Lots of small suggestions

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda (hunk 6 above)
Comment (felixwellen) [reply Cubical/Algebra/CommAlgebra/Instances/Initial.agda:46]: Can I consider this as resolved, now that makeCommAlgebraHom is defined by a copattern?

**Comment by felixwellen** [comment]
Sorry for the mess! I should look closer, when porting old code...

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda (hunk 6 above)
Comment (mortberg) [reply Cubical/Algebra/CommAlgebra/Instances/Initial.agda:46]: @Saizan : Isn't it actually the case that even if the function that one uses is defined using copatterns that is not going to help here and we should use copatterns here as well?

File: Cubical/Algebra/CommAlgebra/Instances/Initial.agda (hunk 7 above)
Comment (felixwellen) [reply Cubical/Algebra/CommAlgebra/Instances/Initial.agda:21]: I made this module anonymous.

**Pull request merged**